### PR TITLE
BN: Use sample variance - unbiased estimator of population variance

### PR DESCRIPTION
**Pull request by ozabluda**
Resolves Issue https://github.com/keras-team/keras/issues/8982

Currently, with `N=2`, the code in Issue https://github.com/keras-team/keras/issues/8982 produces wrong population variance (**0.47084832**), while the correct one is **1.0+-...** (with this PR the result is **0.99781173**). This is due to current code using wrong "variance of the sample"  `1/N` biased estimator instead of correct unbiased "sample variance" estimator `(1/(N-1)`, which for N=2 is exactly 2x smaller.

This is contrary to the "Batch Normalization" paper (https://arxiv.org/abs/1502.03167), where Algorithm 2, step 10 uses unbiased estimator: 
```
Var[x] ← m/(m−1) E_B[σ_B^2]
``` 
and everywhere else the paper consistently uses `Var[x]`, not `E[Var[B]]` or something (although "Batch renormalization" paper has it garbled).

The reason we don't typically see the obvious manifestation of the bug is because typically N is larger, and we normalize across the whole feature map (but we should see is for small N for Dense layers). Nevertheless, for **cifar10_resnet.py** example, after **epoch 1**,  **val_acc** improves from **0.5118** to **0.5235** (and together with PR https://github.com/keras-team/keras/pull/9003 improves again to **0.5860**. FWIW, final **val_acc** "improved" from **0.9195** to **0.9223**).

### CODE REVIEW

**Comment by ozabluda** [comment]
@souptc, @fchollet, does this trigger a bug in CNTK backend (which I don't know how to fix) or I should be doing it differently?

**Comment by fchollet** [comment]
Most of all, we want our BN layer to have the same behavior as the `tf.layers.BatchNormalization` layer. I believe this would no longer be the case?

**Comment by ozabluda** [comment]
au contraire, I think it'll make us consistent with it:
https://github.com/tensorflow/tensorflow/blob/r1.4/tensorflow/python/layers/normalization.py#L309

or not. it's kinda hard to tell:
https://github.com/tensorflow/tensorflow/search?q=Bessel+correction

**Comment by ozabluda** [comment]
If you know how to convert code in Issue https://github.com/keras-team/keras/issues/8982 to raw TF, you can determine it experimentally.

**Comment by ozabluda** [comment]
@fchollet, here is my stab at experimental determination of what `tf.layers.BatchNormalization` does (following https://github.com/keras-team/keras/pull/9028#issuecomment-356488984), but I don't know how to extract population statistics from it, which is not used during training, and I am not sure what this code does for validation (probably runs it in training mode).

Code ideas are taken from https://stackoverflow.com/questions/44991470/using-tensorflow-layers-in-keras

```
from keras.models import Sequential
from keras.layers import BatchNormalization, Lambda, Dense, InputLayer
from keras.optimizers import SGD
import tensorflow as tf
import numpy as np

m = Sequential()
m.add(InputLayer(input_shape=(1,)))
m.add(InputLayer(
    input_tensor=tf.layers.batch_normalization(m.layers[0].output)[0]))

m.summary()
print(m.layers[0].weights)
N=2

x  = np.random.normal(loc=0, scale=1, size=10000)
print(np.var(x, axis=None), np.mean(x, axis=None))
y  = x

m.compile(optimizer='sgd', loss='mse')
print("Before evaluate:", m.evaluate(x,y))
print(np.array(m.layers[0].get_weights()).ravel())

m.fit(x, y, verbose=1, epochs=2, shuffle=False, batch_size=N,
      validation_data=(x,y))

print("After evaluate:", m.evaluate(x,y), m.predict(x))
print(np.array(m.layers[0].get_weights()).ravel())
```

With current code (or this PR, which does not affect training), during training, with all default BN parameters, with `batch_size=2`, Keras loss is `0.8533`, while tf loss is `1.0285`, so there is some discrepancy already. With `batch_size=100`, both Keras and tf loss is close to 0.

**Comment by ozabluda** [comment]
@fchollet, as far as I can tell, this PR makes Keras **consistent** with current fused (default) `tf.layers.BatchNormalization`. Annotated references below:

https://github.com/tensorflow/tensorflow/blob/a1b155b6f51c539974e1bf73d0e0d15b388b9219/tensorflow/python/layers/normalization.py#L400
```
   if not self._bessels_correction_test_only:
      # Remove Bessel's correction to be consistent with non-fused batch norm.
      # Note that the variance computed by fused batch norm is
      # with Bessel's correction.
      sample_size = math_ops.cast(
          array_ops.size(inputs) / array_ops.size(variance), variance.dtype)
      factor = (sample_size - math_ops.cast(1.0, variance.dtype)) / sample_size
      variance *= factor
```
You can see from the comments that fused batch norm is **with** Bessel's correction (same as this PR)

https://github.com/tensorflow/tensorflow/blob/a1b155b6f51c539974e1bf73d0e0d15b388b9219/tensorflow/python/layers/normalization.py#L168
`self._bessels_correction_test_only = True`

Current TF code by default does **not** remove Bessel's correction. That private undocumented hook to remove it (for "test only") was introduced in https://github.com/tensorflow/tensorflow/commit/594f3208377241785bb1451cd2b03fcea1da0a38 to enable a unit test.

Experimentally, the code in the next comment has "variance of the sample" 4.0, but `tf.layers.BatchNormalization` `moving_variance=7.9996977` (i.e. **with** Bessel's correction).

The comment `# Remove Bessel's correction to be consistent with non-fused batch norm.` means that there is internal inconsistency within TF between fused and non-fused BN, but fused is the default.

**Comment by ozabluda** [comment]
Code to illustrate the above comment. Adopted from https://github.com/tensorflow/tensorflow/issues/15250
```
import numpy as np
import tensorflow as tf
from tensorflow.python.ops import control_flow_ops

with tf.Session() as sess:
    image_shape = (2, 1,1,1)
    image_values = [[[[-2]]], [[[+2]]]]
    print(image_values)
    expected_mean = np.mean(image_values)
    expected_var = np.var(image_values)
    print("expected", expected_mean, expected_var)
    images = tf.constant(image_values, shape=image_shape, dtype=tf.float32)
    output = tf.layers.batch_normalization(images, training=True)
    #layer = tf.layers.BatchNormalization(fused=True)
    update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
    assert len(update_ops) == 2
    with tf.control_dependencies(update_ops):
        barrier = tf.no_op(name='barrier')
    output = control_flow_ops.with_dependencies([barrier], output)
    # Initialize all variables
    sess.run(tf.global_variables_initializer())
    moving_mean = tf.contrib.framework.get_variables(
        'batch_normalization/moving_mean')[0]
    moving_variance = tf.contrib.framework.get_variables(
        'batch_normalization/moving_variance')[0]
    mean, variance = sess.run([moving_mean, moving_variance])
    ## After initialization moving_mean == 0 and moving_variance == 1.
    print(mean, variance)
    for _ in range(1000):
        sess.run([output])
    mean = moving_mean.eval()
    variance = moving_variance.eval()
    print(mean, variance)
```

Output:
```
[[[[-2]]], [[[2]]]]
expected 0.0 4.0
[0.] [1.]
[0.] [7.9996977]
```

**Comment by fchollet** [comment]
Ok, sounds good then. Meanwhile, do we have a fix for the CNTK errors?

**Comment by ozabluda** [comment]
I haven't seriously attempted to fix CNTK errors, because I an not sure if it's a bug in CNTK backend or I am doing it wrong. I don't have CNTK, so Travis would be hammered.

Maybe something like this would work better?
https://github.com/keras-team/keras/blob/321d838ba880c3c3529997354617963bb501eaec/examples/deep_dream.py#L90

**Comment by fchollet** [comment]
@souptc would you have any thoughts on how to handle `K.prod(K.shape(inputs))` with CNTK?

**Comment by souptc** [comment]
Sorry for the late reply, I just came back from vacation.

`K.prod(K.shape(inputs))` won't work for cntk because CNTK doesn't support symbolic shape (it actually return int shape in cntk backend), so it will fail for dynamic axis.

Add Emad, who is leading cntk's development now.

@ebarsoumMS, do you have any suggestion about in CNTK, how to calculate sample size in a symbolic way? We do this in cntk's batch normalization api but seems didn't expose that.

**Comment by ozabluda** [comment]
ping, @ebarsoumMS, @souptc

**Comment by ebarsoumMS** [comment]
Hi @ozabluda, sorry for late reply, we have symbolic shape in our backlog but we won't be able to get it in soon. Can you disable CNTK BN for the time being?

**Comment by fchollet** [comment]
> Can you disable CNTK BN for the time being?

Wow, that sounds a bit extreme? Are there no users of the CNTK backend out there?

**Comment by fchollet** [comment]
@ozabluda speaking of BN issues, could you also look into the issues caused by the shadow variables created by `zero_debias=True` recently enabled for TF BN moving average updates? e.g. #9165 #9126

I think a workaround could be to pass a unique name argument to the op in the TF backend (if necessary), or some such.

**Comment by ozabluda** [comment]
I disabled this PR for CNTK, tests are passing now. BN still works for CNTK, just like before this PR. When  symbolic shape works in CNTK, we can re-enable this PR for CNTK.

**Comment by fchollet** [comment]
@ozabluda regarding the `zero_debias` issue, name unification for the shadow variables appears to have been done in TF as of 1.4. So this is already fixed. Never mind.

**Comment by ozabluda** [comment]
For the record, [Group Normalization (2018) Yuxin Wu, Kaiming He](https://arxiv.org/abs/1803.08494):
>it is required for BN to work with a sufficiently large batch size (e.g., 32 per worker. A small batch leads to inaccurate estimation of the batch statistics, and reducing BN’s batch size increases the model error dramatically (Figure 1). 

also see related Issue https://github.com/tensorflow/tensorflow/issues/18222
